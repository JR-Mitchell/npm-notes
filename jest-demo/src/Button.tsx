//Import from external 'react' module
import React from 'react';

class Button extends React.Component<{},{paraText: string, timeout?: ReturnType<typeof setTimeout>}> {
    constructor(props) {
        super(props);
        this.state = {paraText:""}
    }

    componentWillUnmount() {
        let timeout = this.state.timeout;
        if (timeout !== undefined) {
            clearTimeout(timeout);
        }
    }

    render() {
        const paraText = this.state.paraText;

        return <div>
            <button onClick={()=>{this.textCountdown(3);}}>
                Click here!
            </button>
            <p>
                {paraText}
            </p>
        </div>
    }

    textCountdown(secs: number) {
        if (secs > 0) {
            this.setState({
                paraText: "Waiting for "+secs.toString()+" seconds...",
                timeout: setTimeout(()=>{this.textCountdown(secs-1)},1000)
            });
        } else {
            this.setState({
                paraText: "Successfully waited!",
                timeout: undefined
            });
        }
    }
}

export default Button;
