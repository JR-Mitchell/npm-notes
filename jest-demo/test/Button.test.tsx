import React from 'react';
import renderer from 'react-test-renderer';
import Button from '../src/Button';


test("Button snapshot", () => {
    jest.useFakeTimers();
    const button = renderer.create(<Button />);
    let tree = button.toJSON();
    expect(tree).toMatchSnapshot("initial");
    //Click the button
    tree.children[0].props.onClick();
    do {
        //Re-render the tree
        tree = button.toJSON();
        //Match new snapshot
        expect(tree).toMatchSnapshot("afterClick");
        //Run only currently set timers
        jest.runOnlyPendingTimers();
    } while(jest.getTimerCount() != 0);
})
