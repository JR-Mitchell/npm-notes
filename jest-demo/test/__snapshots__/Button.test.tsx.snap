// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot: afterClick 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Click here!
  </button>
  <p>
    Waiting for 3 seconds...
  </p>
</div>
`;

exports[`Button snapshot: afterClick 2`] = `
<div>
  <button
    onClick={[Function]}
  >
    Click here!
  </button>
  <p>
    Waiting for 2 seconds...
  </p>
</div>
`;

exports[`Button snapshot: afterClick 3`] = `
<div>
  <button
    onClick={[Function]}
  >
    Click here!
  </button>
  <p>
    Waiting for 1 seconds...
  </p>
</div>
`;

exports[`Button snapshot: afterClick 4`] = `
<div>
  <button
    onClick={[Function]}
  >
    Click here!
  </button>
  <p>
    Successfully waited!
  </p>
</div>
`;

exports[`Button snapshot: initial 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Click here!
  </button>
  <p>
    
  </p>
</div>
`;
